src/graph/network/communicator.h

```diff
@@ -6,112 +6,144 @@
 #ifndef DGL_GRAPH_NETWORK_COMMUNICATOR_H_
 #define DGL_GRAPH_NETWORK_COMMUNICATOR_H_
 
+#include <dmlc/logging.h>
+
 #include <string>
 
+#include "msg_queue.h"
+
 namespace dgl {
 namespace network {
 
 /*!
  * \brief Network Sender for DGL distributed training.
  *
- * Sender is an abstract class that defines a set of APIs for sending 
- * binary data over network. It can be implemented by different underlying 
- * networking libraries such TCP socket and ZMQ. One Sender can connect to 
- * multiple receivers, and it can send data to specified receiver via receiver's ID.
+ * Sender is an abstract class that defines a set of APIs for sending binary 
+ * data message over network. It can be implemented by different underlying 
+ * networking libraries such TCP socket and MPI. One Sender can connect to 
+ * multiple receivers and it can send data to specified receiver via receiver's ID.
  */
 class Sender {
  public:
+  /*!
+   * \brief Sender constructor
+   * \param queue_size size (bytes) of message queue. 
+   * Note that, the queue_size parameter is optional.
+   */
+  explicit Sender(int64_t queue_size = 0) {
+    CHECK_GE(queue_size, 0);
+    queue_size_ = queue_size;
+  }
+
   virtual ~Sender() {}
 
   /*!
-   * \brief Add receiver address and it's ID to the namebook
-   * \param ip receviver's IP address
-   * \param port receiver's port
+   * \brief Add receiver's address and ID to the sender's namebook
+   * \param addr Networking address, e.g., 'socket://127.0.0.1:50091', 'mpi://0'
    * \param id receiver's ID
    */
-  virtual void AddReceiver(const char* ip, int port, int id) = 0;
+  virtual void AddReceiver(const char* addr, int id) = 0;
 
   /*!
    * \brief Connect with all the Receivers
-   * \return True for sucess and False for fail
+   * \return True for success and False for fail
    */
   virtual bool Connect() = 0;
 
   /*!
-   * \brief Send data to specified Receiver
-   * \param data data buffer for sending
-   * \param size data size for sending
+   * \brief Send data to specified Receiver.
+   * \param msg data message
    * \param recv_id receiver's ID
-   * \return bytes we sent
+   * \return bytes of data
    *   > 0 : bytes we sent
    *   - 1 : error
    */
-  virtual int64_t Send(const char* data, int64_t size, int recv_id) = 0;
+  virtual int64_t Send(Message msg, int recv_id) = 0;
```

Comment:
Please update the comment:
(1) The send is non-blocking. There is no guarantee that the message has been physically sent out when the function returns.
(2) The communicator will assume the responsibility of the given message.
(3) The API is multi-thread safe.
(4) Messages sent to the same receiver are guaranteed to be received in the same order. There is no guarantee for messages sent to different receivers.

Reply:
I will update it in socket_communicator because the behavior of different communicator implementation could be different.

Reply:
> I will update it in socket_communicator because the behavior of different communicator implementation could be different.

For example, SocketComminactor is thread-safe but it's not in MPI implementation.

Reply:
Why MPI implementation cannot be thread-safe?

Reply:
> Why MPI implementation cannot be thread-safe?

If let me implement mpi_communicator, the send() API will call mpi send directly. But I remember that all the mpi send/recv must be invoked in a single thread.

Reply:
You could still implement a thread-safe version by the message queue design. The design of the `Communicator` interface, as you've written in the docstring, is to provide a unified abstraction no matter how it is implemented (mpi or socket). If socket and mpi implementations have different behaviors and the usage will depend on that, does that mean when switching to MPI implementation the codes will be broken? then why need an interface? That's why these behavior definition should be in the interface class.